### PR TITLE
(BIDS-3209) API: notification endpoints

### PR DESCRIPTION
**Pull request by LuccaBitfly**

### CODE REVIEW

File: frontend/types/api/notifications.ts [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+  realtime_mode: boolean;
+}
+export interface NotificationSettingsValidatorDashboard {
+  webhook_url: string;
+  webhook_discord: boolean;
+  real_time_mode: boolean;
+  subscribed_events: NotificationEventsValidatorDashboard;
+}
+export type InternalPutNotificationSettingsValidatorDashboardResponse = ApiDataResponse<NotificationSettingsValidatorDashboard>;
+export interface NotificationEventsAccountDashboard {
```
Comment (thib-wien) [inline frontend/types/api/notifications.ts:151]: Is this the part that tells us how to send the API the preferences of the user in this dialog?
image
The user can select networks and can ask us to ignore spam transactions.
I do not see those fields in this json :)

File: frontend/types/api/notifications.ts [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+  incoming_transactions: boolean;
+  outgoing_transactions: boolean;
+  track_erc20_token_transfers: boolean;
```
Comment (thib-wien) [inline frontend/types/api/notifications.ts:154]: I think here we need a float because the user types a  value in dollars.

**Comment by thib-wien** [comment]
PS: I adopted this convention for numbers (the percentage of offline validators and the threshold for ERC20 transaction) :
- positive number: the user entered a number and activated the setting with the checkbox
- negative number: the user entered a number and deactivated the setting with the checkbox
- NaN : the user has left the input field empty

File: frontend/types/api/notifications.ts (hunk 1 above)
Comment (LuccaBitfly) [reply frontend/types/api/notifications.ts:151]: Network selection is in `NotificationSettingsAccountDashboard`, this type is supposed to only contain subscribable events. But @MauserBitfly would prefer flat types, so I will add all fields from the events types to their respective settings type.

File: frontend/types/api/notifications.ts [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+  enable_email: boolean;
+  enable_push: boolean;
+  subscribed_events: NotificationEventsGeneral;
+  networks: NotificationSettingsNetwork[];
+  paired_devices: NotificationSettingsPairedDevice[];
+}
+export type InternalGetNotificationSettingsResponse = ApiDataResponse<NotificationSettings>;
+export interface NotificationEventsValidatorDashboard {
+  validator_offline: boolean;
+  group_offline: number /* float64 */;
```
Comment (thib-wien) [inline frontend/types/api/notifications.ts:135]: As JSON does not support `NaN`, can you make tygo generate `number|null` here?
`null` is when the user hasn't set a value (the user never opened the setting dialog or has left the input field empty)

File: frontend/types/api/notifications.ts [hunk 4]
```diff
@@ -149,6 +149,7 @@ export interface NotificationSettingsAccountDashboard {
   subscribed_chain_ids: number /* uint64 */[];
   is_incoming_transactions_subscribed: boolean;
   is_outgoing_transactions_subscribed: boolean;
+  is_erc20_token_transfers_subscribed: number /* float64 */;
```
Comment (thib-wien) [inline frontend/types/api/notifications.ts:152]: Maybe here we want a boolean (the next member contains the amount).
Or this member disappears and we use `< 0` to disable

Comment (LuccaBitfly) [reply frontend/types/api/notifications.ts:152]: Yes it should be a boolean, thx

File: backend/pkg/api/handlers/notifications.go [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+	"errors"
+	"math"
+	"net/http"
+
+	"github.com/gobitfly/beaconchain/pkg/api/enums"
+	"github.com/gobitfly/beaconchain/pkg/api/types"
+	"github.com/gorilla/mux"
+)
+
+// TODO move to internal.go before merging
```
Comment (remoterami) [inline backend/pkg/api/handlers/notifications.go:13]: Don't forget

File: backend/pkg/api/handlers/notifications.go [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+		handleErr(w, err)
+		return
+	}
+	checkMinMax(&v, req.MachineStorageUsageThreshold, 0, 1, "machine_storage_usage_threshold")
+	checkMinMax(&v, req.MachineCpuUsageThreshold, 0, 1, "machine_cpu_usage_threshold")
+	checkMinMax(&v, req.MachineMemoryUsageThreshold, 0, 1, "machine_memory_usage_threshold")
+	checkMinMax(&v, req.RocketPoolMaxCollateralThreshold, 0, 1, "rocket_pool_max_collateral_threshold")
+	checkMinMax(&v, req.RocketPoolMinCollateralThreshold, 0, 1, "rocket_pool_min_collateral_threshold")
+	// TODO: check validity of clients
+	err := h.dai.UpdateNotificationSettingsGeneral(r.Context(), userId, req)
```
Comment (remoterami) [inline backend/pkg/api/handlers/notifications.go:234]: Add `v.hasErrors()` check

File: backend/pkg/api/handlers/notifications.go [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+		Name                   string `json:"name,omitempty"`
+		IsNotificationsEnabled bool   `json:"is_notifications_enabled"`
+	}{}
+	if err := v.checkBody(&req, r); err != nil {
+		handleErr(w, err)
+		return
+	}
+	// TODO use a better way to validate the paired device id
+	pairedDeviceId := v.checkRegex(reNonEmpty, mux.Vars(r)["paired_device_id"], "paired_device_id")
+	v.checkNameNotEmpty(req.Name)
```
Comment (remoterami) [inline backend/pkg/api/handlers/notifications.go:290]: Consider using return value of `checkNameNotEmpty` as name

Also add `v.hasErrors()` check

File: backend/pkg/api/handlers/notifications.go [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+		},
+	}
+
+	returnOk(w, response)
+}
+
+func (h *HandlerService) InternalDeleteUserNotificationSettingsPairedDevices(w http.ResponseWriter, r *http.Request) {
+	var v validationError
+	// TODO use a better way to validate the paired device id
+	pairedDeviceId := v.checkRegex(reNonEmpty, mux.Vars(r)["paired_device_id"], "paired_device_id")
```
Comment (remoterami) [inline backend/pkg/api/handlers/notifications.go:311]: `v.hasErrors()` check again - or did you leave them out intentionally?

**Review by remoterami — CHANGES_REQUESTED** [review]

File: backend/pkg/api/handlers/notifications.go (hunk 8 above)
Comment (LuccaBitfly) [reply backend/pkg/api/handlers/notifications.go:311]: No, it was unintentional 😅

File: backend/pkg/api/handlers/notifications.go (hunk 5 above)
Comment (LuccaBitfly) [reply backend/pkg/api/handlers/notifications.go:13]: Will move after approval

**Review by remoterami — APPROVED** [review]
lgtm

**Pull request merged**